v2/googlecloud-to-splunk/src/main/java/com/google/cloud/teleport/v2/templates/GCSToSplunk.java

```diff
@@ -0,0 +1,291 @@
+   * pipeline is finished before returning.
+   *
+   * @param options The execution options.
+   * @return The pipeline result.
+   */
+  public static PipelineResult run(GCSToSplunkOptions options) {
+    Pipeline pipeline = Pipeline.create(options);
+
+    CoderRegistry registry = pipeline.getCoderRegistry();
+    registry.registerCoderForClass(SplunkEvent.class, SplunkEventCoder.of());
+    registry.registerCoderForType(
+        FAILSAFE_ELEMENT_CODER.getEncodedTypeDescriptor(), FAILSAFE_ELEMENT_CODER);
+
+    PCollectionTuple readCsvTuple = pipeline.apply("Read CSV", readFromCsv(options));
+
+    PCollectionTuple failsafeTransformedLines =
+        convertToFailsafeAndMaybeApplyUdf(readCsvTuple, options);
+
+    PCollectionTuple splunkEventTuple = convertToSplunkEvent(failsafeTransformedLines.get(UDF_OUT));
+
+    PCollection<SplunkWriteError> wrappedSplunkWriteErrors =
+        writeToSplunk(splunkEventTuple.get(SPLUNK_EVENT_OUT), options);
+
+    flattenErrorsAndConvertToString(
+            failsafeTransformedLines.get(UDF_ERROR_OUT),
+            splunkEventTuple.get(SPLUNK_EVENT_ERROR_OUT),
+            wrappedSplunkWriteErrors)
+        .apply("Output Errors To GCS", writeErrorsToGCS(options));
+
+    return pipeline.run();
+  }
+
+  static ReadCsv readFromCsv(GCSToSplunkOptions options) {
+    return CsvConverters.ReadCsv.newBuilder()
+        .setCsvFormat(options.getCsvFormat())
+        .setDelimiter(options.getDelimiter())
+        .setHasHeaders(options.getContainsHeaders())
+        .setInputFileSpec(options.getInputFileSpec())
+        .setHeaderTag(CSV_HEADERS)
+        .setLineTag(CSV_LINES)
+        .setFileEncoding(options.getCsvFileEncoding())
+        .build();
+  }
+
+  static PCollectionTuple convertToFailsafeAndMaybeApplyUdf(
+      PCollectionTuple csvLines, GCSToSplunkOptions options) {
+
+    String transformStepName = "Convert To Failsafe Element";
+
+    // In order to avoid generating a graph that makes it look like a UDF was called when none was
+    // intended, simply convert the input to FailsafeElements.
+    if (Strings.isNullOrEmpty(options.getJavascriptTextTransformGcsPath())) {
+      if (!Strings.isNullOrEmpty(options.getJsonSchemaPath())) {
+        transformStepName = "Apply JSON Schema";
+      }
+      return csvLines.apply(
+          transformStepName,
+          CsvConverters.LineToFailsafeJson.newBuilder()
+              .setDelimiter(options.getDelimiter())
+              .setJsonSchemaPath(options.getJsonSchemaPath())
+              .setHeaderTag(CSV_HEADERS)
+              .setLineTag(CSV_LINES)
+              .setUdfOutputTag(UDF_OUT)
+              .setUdfDeadletterTag(UDF_ERROR_OUT)
+              .build());
+    }
+
+    // For testing purposes, we need to do this check before creating the PTransform. Otherwise, we
+    // get a NullPointerException due to the PTransform not returning
+    // a value.
+    if (Strings.isNullOrEmpty(options.getJavascriptTextTransformFunctionName())) {
```

Comment:
Sorry, missed this the last time. It seems that the `CsvConverters.LineToFailsafeJson` will handle all of the dynamic graph construction, so you can probably just call it with the options.

There is one bug [here](https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/901278b958bfc0b154ca6f0a6ec34d6150ce7a78/v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/CsvConverters.java#L295-L305), though, since the check allows empty strings (it shouldn't) and doesn't account for the function name not being set (it should). However, I'd accept that for now, since this isn't production-ready, and fix that in the next PR.